Add RFC for JDBC Join Push down (#32)
prestodb/rfcs: RFC-0009-jdbc-join-push-down.md

## JdbcComputePushdown Optimizer
Using JdbcComputePushdown optimizer, we are pushing down the join criteria as additional predicate. For doing this, we enhanced JdbcComputePushdown optimizer by adding a join predicate to sql translator. 
```
private final JdbcJoinPredicateToSqlTranslator jdbcJoinPredicateToSqlTranslator;
```

```
this.jdbcJoinPredicateToSqlTranslator = new JdbcJoinPredicateToSqlTranslator(
                functionMetadataManager,
                buildFunctionTranslator(ImmutableSet.of(JoinOperatorTranslators.class)),
                identifierQuote);
```
We have also added visitPlan() method and enhanced the visitFilter() method.

Enhancements done to visitFilter() are from this PR : https://github.com/prestodb/presto/pull/16412/

> Before, if all the filter could not be pushed down, nothing would be pushed down.

Reviewer: ```suggestion
Before, if the complete filter could not be pushed down, nothing would be pushed down.
```

The changes help in finding what all can be pushed down and able to be translated. The filters that can't be pushed down are kept in a new FilterNode.

## Changes in SPI module 

> - Added 2 new classes : JoinTableInfo, JoinTableSet

Reviewer: Can you add a basic class outline for `JoinTableInfo` and `JoinTableSet` ? Just the members and any inheritance relationships would suffice

- JoinTableInfo holds the information about a table that is undergoing pushdown. It has details about tableHandle, assignments and outputVariables.
- JoinTableSet is used to store a set of JoinTableInfo objects.
- These are explained in detail below.

We have added a new Connector capability called SUPPORTS_JOIN_PUSHDOWN ConnectorCapabilities enum in SPI. Any connector which wants to participate in pushdown need to added this Connector capability to their Set of ConnectorCapabilities. JDBC has done it like [this](#3-use-multijoinnode-to-group-jdbc-tables-based-on-connector-name)

## Edge cases

### Self Join

> Self join is a scenario where a table is joined with itself. This works natively in Presto. When you do a self join in presto, that is sent as select statements to the datasource first. On top of that presto does all the joining and filtering etc. 

Reviewer: ```suggestion
Self join is a scenario where a table is joined with itself. This is handled in Presto by assigning unique TableScanNode's for each instance of the table being joined
```

nit : Please keep all refs to `Presto` with an uppercase-P since its a proper noun

We want to push down this self join to database level. This is similar to pushing down an inner join. But, we faced some difficulties in achieving this.

#### Difficulty faced :
If we do a join query which joins the same table with itself, We get 2 TableScans in the plan.
There isn’t a way to differentiate between the columns of these TableScans.

Example : Assume there is a column 'col_1' in a table 'table_1'.  We join table_1 with itself ->
```
select * from table_1 a join table_1 b on a.col_1 = b.col_1;
```
The column 'col_1' will be referred as col_1 (in outputVariables) inside TableScan1 and will point to 'col_1' JdbcColumnHandle (in assignments)
It will be referred as col_1_0 (in outputVariables) inside TableScan2 and will point to 'col_1' JdbcColumnHandle (in assignments)

> When our Optimiser 'JdbcJoinRenderByConnector' is run, both these TableScans get converted to a Single TableScan, with assignments, outputVariables, etc. combined.

Reviewer: ```suggestion
When our Optimiser `JdbcJoinRenderByConnector` is run, both these TableScan's get converted to a single TableScan, with assignments, output variables, etc. are combined.
```

(Note : assignments now has two references to the same column 'col_1'. One from 'col_1' and the other from 'col_1_0')
This creates a problem in pushPredicateIntoTableScan method in PickTableLayout class, which assumes that all the values in assignments are unique.
This assumption by presto is done on the basis that till now all queries done by presto to any database is as a select from a single table.

#### Proposed Solution :

We can bring in a 'table alias' to uniquely identify the tables. This is not the table alias that the user passes in his query.
This alias can then be used to unique identify columns and assignments even in the case of self joins.

> This alias can then also be used in the Join conditions, predicates, filters, etc.. which will simplify the Pushdown Query that we create in Query Builder.
> This solved the issue and we were able to push down self joins as well.

Reviewer: ```suggestion
This alias can then be used to build the pushdown filter clauses from the query predicates
```
`conditions, predicates, filters,` all refer to the same thing, so lets make this more precise

Addition of JoinTableInfo Class:

> A new class, JoinTableInfo, has been introduced to encapsulate essential information for each join table. This includes:

Reviewer: If we add a class stub in the `SPI changes` section above, we could refer to the members directly and provide a more concise step-by-step description of how the alias is built

Author: Linked.

Creating the New TableScanNode:
- The updated list of ConnectorTableHandle and updated assignments are then used to construct the new TableScanNode

## Create Join Query in QueryBuilder

At present we are focusing on common operators =, <, >, <=, >= and !=  with common datatype like int, bigint, float, real, string, varchar, char. So there is no connector level implementation required and focusing on single implementation for all supported Jdbc connector through QueryBuilder class.

Now we have new TableScanNode with list of joinTables. The split manager works on JdbcSplit and provide objects for processing on connector level. We need to modify the logic to transfer the new object (Optional<List<ConnectorTableHandle>> joinPushdownTables) too, to the connector level. Currently, for the Join query, the split is transfer to buildSql() to create the select statement. If the split contains ‘joinTables’ details then we need to transfer this ‘joinTables’ details to the new method called ‘buildJoinSql()’ where we need to create and return join query instead of select query.

> On buildJoinSql(), we need to handle columns to be selected, the tables from which to query, the join condition and filter conditions if any.

Author: Point 1 and 2 are done.

Point 3, about adding a picture - I will add it.

> On buildJoinSql(), we need to handle columns to be selected, the tables from which to query, the join condition and filter conditions if any.

Author: Added.

Not Applicable

- If we need special migration tools, describe them here.

Not Applicable

- When will we remove the existing behaviour, if applicable.

Not Applicable

- How should this feature be taught to new and existing users? Basically mention if documentation changes/new blog are needed?

Yes, documentation changes will be required. 

- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?

1. Push down all kinds of Joins
2. Some queries may become slower due to this change, if the session flag is set to true.

## Test Plan

Added 3 new test classes - TestJdbcJoinPushdown, TestJdbcQueryBuilderJoinPushdown, TestJdbcQueryBuilderJoinPushdownExtended
Added 1 new class TestJoinQueriesWithPushDown which extends AbstractTestJoinQueries. 
Total 221 new test cases - all are passing

All the test cases in AbstractTestJoinQueries are passing with the join pushdown flag enabled. This gives us a lot of confidence on the correctness of the implementation.

We have done a POC on the implementation and we were able to see following performance improvements : 

> ![Performance Image 1](RFC-0009-jdbc-join-push-down/perf_1.png)

Reviewer: 1. Can you make this a markdown table instead of this image ?
2. To this table, can you add two more columns - first column is the output of `EXPLAIN ANALYZE` with the feature turned off, the second one is the same with it turned on. Since the text of the `EXPLAIN ANALYZE` is large, you can attach the outputs and link to them

Author: Made it into a table.. Attached query and explain analyze as txt files.. But need to fill the txt files with the actual outputs. Will do that.

Author: Added.

# **RFC0009-jdbc-join-push-down for Presto**

## Jdbc join push down in presto

Proposers

* Ajas M M
* Haritha K
* Thanzeel Hassan
* Glerin Pinhero

## Related Issues

https://github.com/prestodb/presto/issues/23152

## Summary

At present, when a query joins multiple tables, it creates a separate TableScanNode for each table. Each TableScanNode select all the records from that table. The join operation is then executed in-memory in Presto using a JOIN node by applying JoinCriteria, FilterPredicate and other criteria (like order by, limit, etc.).

> However, if the query joins tables from the same JDBC datasource, it would be more efficient to let the datasource handle the join instead of creating a separate TableScanNode for each table and joining them in Presto. If we "Push down" or send these joins to remote JDBC datasource it increases the query performance. i.e., decreases the query execution time. We have seen improvements from 3x to 10x.

Author: I have added a table which shows the data i got from the presto UI. I dont think it shows Disk and network usage.. 
Can you take a look ? @aaneja

> However, if the query joins tables from the same JDBC datasource, it would be more efficient to let the datasource handle the join instead of creating a separate TableScanNode for each table and joining them in Presto. If we "Push down" or send these joins to remote JDBC datasource it increases the query performance. i.e., decreases the query execution time. We have seen improvements from 3x to 10x.

Reviewer: The table LGTM. The `EXPLAIN ANALYZE` output has sufficient details to infer network usage

    - RemoteStreamingExchange[PlanNodeId 233][GATHER] => [order_id:integer, c_customer_id:char(16)]
            Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: 0.00, network: ?}
        - TableScan[PlanNodeId 217][TableHandle {connectorId='postgresql', connectorHandle='postgresql:public.orders:null:public:orders', layout='Optional[{domains=ALL, additionalPredicate={}}]'}] => [order_id:integer, c_customer_id:char(16)]
                Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: 0.00, network: 0.00}
                LAYOUT: {domains=ALL, additionalPredicate={}}
                order_id := JdbcColumnHandle{connectorId=postgresql, columnName=order_id, jdbcTypeHandle=JdbcTypeHandle{jdbcType=4, jdbcTypeName=int4, columnSize=10, decimalDigits=0, arrayDimensions=null}, columnType=integer, nullable=true, comment=Optional.empty} (1:45)
                c_customer_id := JdbcColumnHandle{connectorId=postgresql, columnName=c_customer_id, jdbcTypeHandle=JdbcTypeHandle{jdbcType=1, jdbcTypeName=bpchar, columnSize=16, decimalDigits=0, arrayDimensions=null}, columnType=char(16), nullable=true, comment=Optional.empty} (2:12)
``` 

> **Original presto plan performance**

Author: Added Explain Analyze outputs

![Original presto plan performance](RFC-0009-jdbc-join-push-down/1_perf_wxd.png)

**Joinpushdown presto plan performance**

![Joinpushdown presto plan performance](RFC-0009-jdbc-join-push-down/2_perf_joinwxd.png)  

## Background

> This implementation is to address a performance limitation of Presto federation of SQLs of JDBC connectors to remote data sources such as DB2, Postgres, Oracle etc. Currently, Presto support predicate pushdown (WHERE condition pushdown) to some extent in JDBC connectors, but it does not have any join pushdown capabilities. This causes high performance impact on join queries and it is raised by some of our clients.

Author: removed

## Proposed Implementation

> At present, if presto get a join query (from the CLI or UI) which is trying to join tables either from same datasource or from different datasource, it is received as a string formatted sql query. Presto validates the syntax and converts it to Query (Statement) object using presto parser and analyzer. This Query object is converted to presto internal reference architecture called Plan, using its logical and physical optimizers. Finally, this plan is executed by the executor.

Author: Maybe I can pick that up post this RFC.

![Joinpushdown presto plan performance](RFC-0009-jdbc-join-push-down/basic_wrk.png)  

> Basically the PlanNode or the Plan is a tree datastructure which represents the sql query. When a join Query is received in logical planning, presto creates a PlanNode with a JoinNode. JoinNode is a tree structure which can hold another node, left table, right table, join conditions, projections and filters related to that join query. If there are multiple tables to join then it create a join tree structure where the left side of the JoinNode will be another JoinNode which hold sub joins to resolve multiple tables. The logical PlanNode is created in such a way, where the first table (the first table in the from clause) is resolved first either from the left TableScanNode or from JoinNode hierarchy using left dept first algorithm, then its adjacent table (very next table) as right side.  So the order and position of the tables in the join query plays an important role to determine query pushdown. 

Author: Reworded with the above statements

JOIN db2.db2.mydb2_table3 t5 ON t4.dbfourthtablecolumn = t5.dbfifthtablecolumn
``` 

**Here we have five tables,** 

mypg_table1 and  mypg_table2 from postgresql connector

mydb2_table1, mydb2_table2 and mydb2_table3 from db2 connector

At present, presto creates five select statement (TableScanNodes) for this query as follows

| No | Node Description                     | SQL Query                                                         |
|----|--------------------------------------|-------------------------------------------------------------------|
| 1  | TableScanNode for mypg_table1        | `select t * from postgresql.pg.mypg_table1 t1`                    |
| 2  | TableScanNode for mypg_table2        | `select t * from postgresql.pg.mypg_table2 t2`                    |
| 3  | TableScanNode for mydb2_table1       | `select t * from db2.db2.mydb2_table1 t3`                         |
| 4  | TableScanNode for mydb2_table2       | `select t * from db2.db2.mydb2_table2 t4`                         |
| 5  | TableScanNode for mydb2_table3       | `select t * from db2.db2.mydb2_table3 t5`                         |

> 

Author: Changed

> In our proposed implementation, we restrict select statement (TableScanNode) creation based on tables and instead create select statement (TableScanNode) against each connector by grouping the tables based on the connector

Author: Added a section heading and reworded.

> | No | Node Description                                  | SQL Query                                                                                                  |

Author: done

|----|---------------------------------------------------|------------------------------------------------------------------------------------------------------------|
| 1  | TableScanNode [mypg_table1, mypg_table2] for PostgreSQL | `select * from postgresql.pg.mypg_table1 t1, postgresql.pg.mypg_table2 t2 where t1.pgfirsttablecolumn=t2.pgsecondtablecolumn` |
| 2  | TableScanNode [mydb2_table1, mydb2_table2, mydb2_table3] for DB2 | `select * from db2.db2.mydb2_table1 t3, db2.db2.mydb2_table2 t4, db2.db2.mydb2_table5 t5 where t3.dbthirdtablecolumn = t4.dbfourthtablecolumn and t4.dbfourthtablecolumn = t5.dbfifthtablecolumn` |

For performing this jdbc join pushdown,  we need to create two logical optimizers GroupInnerJoinsByConnector and JdbcJoinPushdown. 

GroupInnerJoinsByConnector optimizer is a PlanOptimizer which is responsible for flattening the JoinNode and adding the sundered nodes to a data structure called MultiJoinNode. 

> GroupInnerJoinsByConnector optimizer will work on MultiJoinNode and will group TableScanNodes based on connector name if the connector supports join pushdown. This optimizer will create a single TableScanNode by using a new data structure called JoinTableSet from the grouped TableScanNode. JoinTableSet is a set of ConnectorTableHandles which is generated from grouped TableScanNode. This optimizer also creates a combined overall predicate and overall assignments for the JoinTableSet and will add these to the newly created TableScanNode. This newly created TableScanNode structure will replace the source list of MultiJoinNode.

Author: changed.

> GroupInnerJoinsByConnector optimizer will then work on re-creating join node with updated MultiJoinNode structure. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#groupinnerjoinsbyconnector-optimizer)

Author: That link was what was copied when I clicked copy to clipboard in the github UI.

Fixed now.

Sql Query : 
``` 
select t1.intcolumn2
from postgresql.pg.mypg_table1 t1
join postgresql.pg.mypg_table2 t2 on t1.pgfirsttablecolumn = t2.pgsecondtablecolumn
Join db2.db2.mydb2_table1 t3 on t3.dbthirdtablecolumn = t2.pgsecondtablecolumn
JOIN db2.db2.mydb2_table2 t4 ON t3.dbthirdtablecolumn = t4.dbfourthtablecolumn
JOIN db2.db2.mydb2_table3 t5 ON t4.dbfourthtablecolumn = t5.dbfifthtablecolumn
``` 
![GroupInnerJoinsByConnector optimizer](RFC-0009-jdbc-join-push-down/after_GroupInnerJoinsByConnector.png)

> JdbcJoinPushdown optimizer is a ConnectorPlanOptimizer, specific to jdbc tables and it generate a single JdbcTableHandle from the grouped ConnectorTableHandle. The low level design is available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#2-jdbcjoinpushdown-optimizer)

Reviewer: Can you fix your links (this and others) to be to the local doc instead of to your private repo ?

Author: Fixed now.

After GroupInnerJoinsByConnector optimizer and JdbcJoinPushdown optimizer, we will invoke existing Predicatepushdown optimizer. PredicatePushdown optimizer will pushdown the filter and join criteria to the re-created JoinNode using the overall predicate and overall assignment. 

After Predicatepushdown optimizer the flow will invoke existing JdbcComputePushdown optimizer and it will pushdown the overall join criteria to the additional predicates.

> After all optimization the PlanNode will pass to the presto-base-jdbc module to create the final join query. The final join query is prepared at the connector level using the Querybuilder. It is explained in the low level design [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#4-create-join-query-in-querybuilder).

Author: Added a diagram. Changed the structure.

A numbered list is provided below in the low level design section. It may be better to remove this from here, and the diagram can be added there. 

Let me check and will update it.

> ## Join query pushdown in presto Jdbc datasource

Author: changed.

|----|-------------------------------------|-------------------------------------------------------------------|
| 1  | TinyINT        | `=, <, >, <=, >=, !=, <>`                    |
| 2  | SmallINT       | `=, <, >, <=, >=, !=, <>`                    |   
| 3  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
| 4  | BigINT         | `=, <, >, <=, >=, !=, <>`                    |
| 5  | Boolean        | `=, !=, <>`                                  |   
| 6  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
| 7  | Real           | `=, <, >, <=, >=, !=, <>`                    |   
| 8  | Double         | `=, <, >, <=, >=, !=, <>`                    |
| 9  | Decimal        | `=, <, >, <=, >=, !=, <>`                    |
| 10 | Varchar        | `=, <, >, <=, >=, !=, <>`                    |
| 11 | Char           | `=, <, >, <=, >=, !=, <>`                    |

**4. All tables from same connector will be grouped based on above specifications and pushed down to underlying datasource.**

**5. Enable presto Join pushdown capabilities by setting the session flag optimizer_inner_join_pushdown_enabled = true.**

> ## Low level Design

Reviewer: Please rework this section or remove it. A lot of this is implementation detail. What we should be highlighting is 
1. How we would group tables by connector using the same mechanism in ReorderJoins. Describe how we would limit the grouping to only join sources that are either `TableScan` or `Filter<-TableScan` or `Project<-Filter<-TableScan` 
2. Describe the updates to the SPI to add a table handle or other data structure to represent a 'group' of individual table handles that are joined
3. Describe how the connectors, and how the JDBC connector in particular can indicate that it wants to 'participate' in this creation of grouped-tables
4. Describe how the JDBC connector will un-wrap these grouped-tables and build the SQL necessary to achieve the JOIN condition that it would use when querying the JDBC source
5. Describe what the corner cases are - why we would need a way to add a table alias

Please create smaller sections for these as you see fit

Reviewer: Additionally, please add a section dedicated to SPI changes - you can add classes, interfaces and types that are being added or removed, and add code-comments on them to describe how they would be used

Author: Removed some things in that section.. Working on the rest

Author: Added the changes in SPI module.

Author: 1. To be added
2. This is added in the SPI module section
3. To be added
4. Described in 'Create Join Query in QueryBuilder' section
5. Made a new section and put self join under it.
